offsetlog.go

```diff
@@ -0,0 +1,197 @@
+	// Dialer used by the offset log to establish connections to the Kafka
+	// brokers. If nil, DefaultDialer is used.
+	dialer *Dialer
+}
+
+const (
+	defaultOffsetLogReadTimeout  = 3 * time.Second
+	defaultOffsetLogWriteTimeout = 3 * time.Second
+	defaultOffsetLogMaxBytes     = 8 * 1024 * 1024
+)
+
+// readOffsets satisfies the offsetStore interface, it returns a list of offsets
+// in the partition read by the log.
+func (log *offsetLog) readOffsets() ([]offset, error) {
+	c, err := log.dialLeader()
+	if err != nil {
+		return nil, err
+	}
+	defer c.Close()
+
+	readTimeout := log.readTimeout
+	if readTimeout <= 0 {
+		readTimeout = defaultOffsetLogReadTimeout
+	}
+	c.SetReadDeadline(time.Now().Add(readTimeout))
+
+	first, last, err := c.ReadOffsets()
+	if err != nil {
+		return nil, err
+	}
+
+	cursor, err := c.Seek(first, SeekAbsolute)
+	if err != nil {
+		return nil, err
+	}
+
+	maxBytes := log.maxBytes
+	if maxBytes <= 0 {
+		maxBytes = defaultOffsetLogMaxBytes
+	}
+	replay := make(map[offsetKey]offsetValue)
+
+	for cursor < last {
+		c.SetReadDeadline(time.Now().Add(readTimeout))
+		b := c.ReadBatch(8, maxBytes)
+
+		for {
+			m, err := b.ReadMessage()
+			if err != nil {
+				break
+			}
+
+			if cursor = m.Offset; cursor < last {
+				var key offsetKey
+				var value offsetValue
+
+				if len(m.Key) != 0 {
+					key.readFrom(m.Key)
+				}
+
+				if len(m.Value) != 0 {
+					value.readFrom(m.Value)
+					replay[key] = value
+				} else {
+					delete(replay, key)
+				}
+			}
+		}
+
+		if err := b.Close(); err != nil {
+			return nil, err
+		}
+
+		cursor++
+	}
+
+	offsets := make([]offset, 0, len(replay))
+	for key, value := range replay {
+		offsets = append(offsets, makeOffsetFromKeyAndValue(key, value))
+	}
+	return offsets, nil
+}
+
+// writeOffsets satisfies the offsetStore interface, it writes a set of offsets
+// to the Kafka partition managed by the log.
+func (log *offsetLog) writeOffsets(offsets ...offset) error {
+	return log.write(makeWriteOffsetsMessages(offsets))
+}
+
+// deleteOffsets satisfies the offsetStore interface, it deletes the offsets from
+// the Kafka partition managed by the log.
+func (log *offsetLog) deleteOffsets(offsets ...offset) error {
+	return log.write(makeDeleteOffsetsMessages(offsets))
+}
+
+func (log *offsetLog) write(msgs []Message) error {
+	c, err := log.dialLeader()
+	if err != nil {
+		return err
+	}
+	defer c.Close()
+	writeTimeout := log.writeTimeout
+	if writeTimeout <= 0 {
+		writeTimeout = defaultOffsetLogWriteTimeout
+	}
+	c.SetWriteDeadline(time.Now().Add(writeTimeout))
+	_, err = c.WriteMessages(msgs...)
+	return err
+}
+
+func (log *offsetLog) dialLeader() (*Conn, error) {
+	if len(log.brokers) == 0 {
+		return nil, errors.New("no brokers configured on the offset log")
+	}
+	dialer := log.dialer
+	if dialer == nil {
+		dialer = DefaultDialer
+	}
+	broker := log.brokers[rand.Intn(len(log.brokers))]
+	return dialer.DialLeader(context.Background(), "tcp", broker, log.topic, log.partition)
```

Comment:
Ideally would we retry on a different broker, if the random broker you fetched from the bootstrap did not connect, until all were exhausted?